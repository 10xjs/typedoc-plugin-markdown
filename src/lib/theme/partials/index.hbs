{{#if groups}}

{{heading 3}} Index

{{#each groups}}

{{heading 4}} {{title}}

{{#each children}}
* [{{{name}}}]({{relativeUrl url}})
{{/each}}

{{/each}}

{{/if}}