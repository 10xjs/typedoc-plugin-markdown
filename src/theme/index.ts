import * as fs from 'fs';
import * as path from 'path';

import { DeclarationReflection, ProjectReflection, Reflection, ReflectionKind } from 'typedoc/dist/lib/models/reflections/index';
import { UrlMapping } from 'typedoc/dist/lib/output/models/UrlMapping';
import { Renderer } from 'typedoc/dist/lib/output/renderer';
import { DefaultTheme } from 'typedoc/dist/lib/output/themes/DefaultTheme';
import { ThemeService } from './service';

export class MarkdownTheme extends DefaultTheme {

  /**
   * This is mostly a copy of the DefaultTheme method with .html ext switched to .md
   * Builds the url for the the given reflection and all of its children.
   *
   * @param reflection  The reflection the url should be created for.
   * @param urls The array the url should be appended to.
   * @returns The altered urls array.
   */
  public static buildUrls(reflection: DeclarationReflection, urls: UrlMapping[]): UrlMapping[] {

    const mapping = DefaultTheme.getMapping(reflection);

    if (mapping) {
      if (!reflection.url || !DefaultTheme.URL_PREFIX.test(reflection.url)) {
        const url = [mapping.directory, DefaultTheme.getUrl(reflection) + '.md'].join('/');
        urls.push(new UrlMapping(url, reflection, mapping.template));
        reflection.url = url;
        reflection.hasOwnDocument = true;
      }
      for (const key in reflection.children) {
        if (reflection.children.hasOwnProperty(key)) {
          const child = reflection.children[key];
          if (mapping.isLeaf) {
            MarkdownTheme.applyAnchorUrl(child, reflection);
          } else {
            MarkdownTheme.buildUrls(child, urls);
          }
        }
      }
    } else {
      MarkdownTheme.applyAnchorUrl(reflection, reflection.parent);
    }

    return urls;
  }

  /**
   * Similar to DefaultTheme method with added functionality to cater for bitbucket heading and single file anchors
   * Generate an anchor url for the given reflection and all of its children.
   *
   * @param reflection  The reflection an anchor url should be created for.
   * @param container   The nearest reflection having an own document.
   */
  public static applyAnchorUrl(reflection: Reflection, container: Reflection) {

    const options = ThemeService.getOptions();

    if (!reflection.url || !DefaultTheme.URL_PREFIX.test(reflection.url)) {

      let anchor = DefaultTheme.getUrl(reflection, container, '.');

      if (reflection['isStatic']) {
        anchor = 'static-' + anchor;
      }

      let anchorRef = '';

      switch (reflection.kind) {
        case ReflectionKind.ExternalModule:
          anchorRef = `external-module-${ThemeService.getAnchorRef(reflection.name)}-`;
          break;
        case ReflectionKind.Class:
          anchorRef = `class-${ThemeService.getAnchorRef(reflection.name)}`;
          break;
        case ReflectionKind.Interface:
          anchorRef = `interface-${ThemeService.getAnchorRef(reflection.name)}`;
          break;
        case ReflectionKind.Module:
          anchorRef = `module-${ThemeService.getAnchorRef(reflection.name)}`;
        case ReflectionKind.Enum:
          const isModule = reflection.parent.kind === 0 || reflection.parent.kind === ReflectionKind.ExternalModule;
          anchorRef = isModule ?
            `module-${ThemeService.getAnchorRef(reflection.name)}` :
            `enumeration-${ThemeService.getAnchorRef(reflection.name)}`;
          break;
        default:
          if (options.mdFlavour === 'bitbucket') {
            let anchorPrefix = '';
            if (reflection.kind === ReflectionKind.ObjectLiteral) {
              anchorPrefix += 'object-literal-';
            }
            reflection.flags.forEach((flag) => {
              anchorPrefix += `${flag}-`;
            });
            const prefixRef = ThemeService.getAnchorRef(anchorPrefix);
            const reflectionRef = ThemeService.getAnchorRef(reflection.name);
            anchorRef = `markdown-header-${prefixRef}${reflectionRef}`;
          } else {
            anchorRef = anchor;
          }

      }

      reflection.url = (container.url !== undefined ? container.url : '') + '#' + anchorRef;
      reflection.anchor = anchor;
      reflection.hasOwnDocument = false;

    }

    reflection.traverse((child: any) => {
      if (child instanceof DeclarationReflection) {
        MarkdownTheme.applyAnchorUrl(child, container);
      }
    });
  }

  constructor(renderer: Renderer, basePath: string, options: any) {
    super(renderer, basePath);

    // remove uneccessary plugins
    renderer.removeComponent('assets');
    renderer.removeComponent('javascript-index');
    renderer.removeComponent('navigation');
    renderer.removeComponent('toc');
    renderer.removeComponent('pretty-print');

    // assign global theme service props
    ThemeService.options = options;
    ThemeService.resources = this.resources;

  }

  /**
   * Test whether the given path contains a documentation generated by this theme.
   *
   * @param path  The path of the directory that should be tested.
   * @returns     TRUE if the given path seems to be a previous output directory,
   *              otherwise FALSE.
   */
  public isOutputDirectory(outPath: string): boolean {
    const files = fs.readdirSync(outPath);
    return fs.existsSync(path.join(outPath, 'README.md')) || (files.length === 1 && path.extname(files[0]) === '.md');
  }

  /**
   * Map the models of the given project to the desired output files.
   *
   * @param project  The project whose urls should be generated.
   * @returns        A list of [[UrlMapping]] instances defining which models
   *                 should be rendered to which files.
   */
  public getUrls(project: ProjectReflection): UrlMapping[] {

    const urls: UrlMapping[] = [];
    const entryPoint = this.getEntryPoint(project);

    ThemeService.projectName = entryPoint.name;

    // pass in additional context
    const additionalContext = {
      displayReadme: this.application.options.getValue('readme') !== 'none',
      hideBreadcrumbs: true,
      isIndex: true,
    };

    const context = { ...entryPoint, ...additionalContext };

    urls.push(new UrlMapping('README.md', context, 'reflection.hbs'));

    if (entryPoint.children) {
      entryPoint.children.forEach((child: DeclarationReflection) => {
        MarkdownTheme.buildUrls(child, urls);
      });
    }

    return urls;
  }

}
