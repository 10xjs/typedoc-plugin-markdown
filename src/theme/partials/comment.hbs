
{{#with comment~}}
{{#if hasVisibleComponent~}}
{{#if shortText~}}{{{ shortText}}}{{/if}}
{{#if text~}}{{{ text}}}{{/if}}
{{#if tags~}}{{~#each tags~}}{{tagName}}{{{text}}}{{/each}}{{/if}}
{{/if}}
{{/with}}
