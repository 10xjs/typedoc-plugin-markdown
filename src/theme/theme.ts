/**
 * Theme definition
 */

/**
 * Typedoc imports
 */
import { DeclarationReflection, ProjectReflection, Reflection } from 'typedoc/dist/lib/models/reflections/index';
import { UrlMapping } from 'typedoc/dist/lib/output/models/UrlMapping';
import { Renderer } from 'typedoc/dist/lib/output/renderer';
import { DefaultTheme } from 'typedoc/dist/lib/output/themes/DefaultTheme';
import { Options } from './options';

interface IOptions {
  markdownFlavour: string;
  markdownSourcefilePrefix: string;
}

export class MarkdownTheme extends DefaultTheme {

  public static buildUrls(reflection: DeclarationReflection, urls: UrlMapping[]): UrlMapping[] {
    const mapping = DefaultTheme.getMapping(reflection);
    if (mapping) {
      const url = [mapping.directory, MarkdownTheme.getUrl(reflection) + '.md'].join('/');
      urls.push(new UrlMapping(url, reflection, mapping.template));
      reflection.url = url;
      reflection.hasOwnDocument = true;
      for (const key in reflection.children) {
        if (reflection.children.hasOwnProperty(key)) {
          const child = reflection.children[key];
          if (mapping.isLeaf) {
            DefaultTheme.applyAnchorUrl(child, reflection);
          } else {
            MarkdownTheme.buildUrls(child, urls);
          }
        }
      }
    } else {
      DefaultTheme.applyAnchorUrl(reflection, reflection.parent);
    }
    return urls;
  }

  public static getUrl(reflection: Reflection, relative?: Reflection, separator: string = '.'): string {

    let url = reflection.getAlias();

    if (reflection.parent && reflection.parent !== relative && !(reflection.parent instanceof ProjectReflection)) {
      url = MarkdownTheme.getUrl(reflection.parent, relative, separator) + separator + url;
    }

    return url;
  }

  private options: IOptions;

  constructor(renderer: Renderer, basePath: string, options: IOptions) {
    super(renderer, basePath);

    this.options = options;

    Options.markdownFlavour = options.markdownFlavour || 'github';
    Options.markdownSourcefilePrefix = options.markdownSourcefilePrefix;

    // remove uneccessary plugins
    renderer.removeComponent('assets');
    renderer.removeComponent('javascript-index');
    renderer.removeComponent('navigation');
    renderer.removeComponent('toc');
    renderer.removeComponent('pretty-print');

  }

  public isOutputDirectory(path: string): boolean {
    return true;
  }

  public getUrls(project: ProjectReflection): UrlMapping[] {
    const urls: UrlMapping[] = [];
    const entryPoint = this.getEntryPoint(project);
    const additionalContext = {
      displayReadme: this.application.options.getValue('readme') !== 'none',
      hideBack: true,
      options: this.options,
    };
    const context = Object.assign(entryPoint, additionalContext);
    // if (this.options.markdownSinglePage) {
    //  urls.push(new UrlMapping('index.md', context, 'reflection.hbs'));
    // } else {
    urls.push(new UrlMapping('index.md', context, 'reflection.hbs'));
    if (entryPoint.children) {
        entryPoint.children.forEach((child: DeclarationReflection) => {
          MarkdownTheme.buildUrls(child, urls);
        });
      // }
    }
    return urls;
  }

}
