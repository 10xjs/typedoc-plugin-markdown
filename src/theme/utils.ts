import * as fs from 'fs';
import * as Handlebars from 'handlebars';
import * as path from 'path';
import { DeclarationReflection, Reflection } from 'typedoc/dist/lib/models/reflections/index';

export function compileTemplate(template: string, data: {}) {
  const source = templateToString(template);
  const md: HandlebarsTemplateDelegate = Handlebars.compile(source);
  return new Handlebars.SafeString(md(data));
}

export function safeString(data: string) {
  return new Handlebars.SafeString(data);
}

export function getAnchorRef(ref: string) {
    return ref.replace(/_|\/|\.| /g, '-').replace(/"/g, '').replace(/ /g, '-').toLowerCase();
  }

function templateToString(template: string) {
  return fs.readFileSync(path.join(__dirname, `/${template}`)).toString();
}
