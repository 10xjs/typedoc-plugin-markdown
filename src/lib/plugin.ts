import * as path from 'path';
import { Application, DeclarationReflection, ProjectReflection } from 'typedoc';
import { Converter } from 'typedoc/dist/lib/converter';
import { Component, ConverterComponent } from 'typedoc/dist/lib/converter/components';
import { Reflection } from 'typedoc/dist/lib/models/reflections/abstract';
import { PageEvent, RendererEvent } from 'typedoc/dist/lib/output/events';
import { Renderer } from 'typedoc/dist/lib/output/renderer';
import { Options } from 'typedoc/dist/lib/utils/options';

import { MarkdownTheme } from './theme/theme';
import { BitbucketTheme } from './theme/theme.bitbucket';
import { DocusaurusTheme } from './theme/theme.docusaurus';
import { GitbookTheme } from './theme/theme.gitbook';

@Component({ name: 'markdown' })
export class MarkdownPlugin extends ConverterComponent {
  static theme: MarkdownTheme | DocusaurusTheme;
  static disableOutputCheck: boolean;
  static application: Application;
  static reflection: Reflection;
  static location: string;
  static page: PageEvent;
  static settings: {
    platform?: 'gitbook' | 'ducusaurus' | 'bitbucket';
    hideSources?: boolean;
    readme?: string;
    includes?: string;
    media?: string;
    mode?: string;
    mdHideSources?: boolean;
    mdSourceRepo?: string;
  };
  static project: ProjectReflection;

  initialize() {
    this.listenTo(this.owner, {
      [Converter.EVENT_RESOLVE_BEGIN]: this.onBegin,
    });
    this.listenTo(this.application.renderer, {
      [PageEvent.BEGIN]: this.onPageBegin,
      [PageEvent.END]: this.onPageEnd,
      [RendererEvent.BEGIN]: this.onRenderBegin,
    });
  }

  onBegin() {
    MarkdownPlugin.application = this.application;
    MarkdownPlugin.settings = this.application.options.getRawValues();
    MarkdownPlugin.setTheme(this.application.renderer, this.application.options);
  }

  onRenderBegin(renderer: RendererEvent) {
    if (MarkdownPlugin.theme) {
      MarkdownPlugin.project = renderer.project;
    }
  }

  onPageBegin(page: PageEvent) {
    if (MarkdownPlugin.theme) {
      MarkdownPlugin.reflection = page.model instanceof DeclarationReflection ? page.model : undefined;
      MarkdownPlugin.location = page.model.url ? page.model.url : '';
    }
  }

  onPageEnd(page: PageEvent) {
    if (MarkdownPlugin.theme) {
      page.contents = page.contents ? MarkdownPlugin.formatContents(page.contents) : '';
    }
  }

  static setTheme(renderer: Renderer, options: Options) {
    const themeName = options.getValue('theme');
    const themePath = path.join(__dirname, './theme/');
    const media = options.getValue('media');
    const platform = options.getValue('platform') || options.getValue('mdEngine');
    const theme = themeName === 'markdown' ? this.getTheme(platform, renderer, themePath, options) : null;
    if (theme) {
      if (media && (theme instanceof DocusaurusTheme || theme instanceof GitbookTheme)) {
        MarkdownPlugin.application.logger.warn(`[typedoc-markdown-plugin] media option is currently not supported in ${platform} theme`);
        options.setValue('media', null);
      }
      renderer.theme = renderer.addComponent('theme', theme);
      MarkdownPlugin.theme = theme;
    } else {
      MarkdownPlugin.application.logger.write('Markdown theme not set');
    }
  }

  static getTheme(platform: string, renderer: Renderer, themePath: string, options: any) {
    if (platform) {
      if (platform === 'gitbook') {
        return new GitbookTheme(renderer, themePath, options);
      }
      if (platform === 'docusaurus') {
        return new DocusaurusTheme(renderer, themePath, options);
      }
      if (platform === 'bitbucket') {
        return new BitbucketTheme(renderer, themePath, options);
      }
    }
    return new MarkdownTheme(renderer, themePath, options);
  }

  static formatContents(contents: string) {
    return contents
      .replace(/^ +/gm, '')
      .replace(/[\r\n]{3,}/g, '\n\n')
      .replace(/!spaces/g, '')
      .trim();
  }
}
