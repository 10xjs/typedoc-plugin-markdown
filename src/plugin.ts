/**
 * Plugin definition
 */
import { Component, ConverterComponent } from 'typedoc/dist/lib/converter/components';
import { Context } from 'typedoc/dist/lib/converter/context';
import { Converter } from 'typedoc/dist/lib/converter/converter';
import { Reflection } from 'typedoc/dist/lib/models/reflections/abstract';
import { Options, OptionsReadMode } from 'typedoc/dist/lib/utils/options';
import { MarkdownTheme } from './theme/theme';

import * as path from 'path';

@Component({ name: 'markdown' })
export class MarkdownPlugin extends ConverterComponent {

  private isSinglePage: boolean;

  public initialize() {

    const options: Options = this.application.options;
    options.read({}, OptionsReadMode.Prefetch);
    this.isSinglePage = (options.getValue('markdownSinglePage'));

    this.listenTo(this.owner, {
      [Converter.EVENT_RESOLVE_BEGIN]: this.onBegin,
    });
  }

  private getThemeDirectory() {
    return path.join(__dirname, '../dist/theme/');
  }

  private onBegin(context: Context, reflection: Reflection) {

    const renderer = this.application.renderer;
    const theme = this.application.options.getValue('theme');
    const themePath = this.getThemeDirectory();
    const options = { isSinglePage: this.isSinglePage };

    // apply the theme
    if (theme === 'markdown') {
      renderer.theme = renderer.addComponent('theme', new MarkdownTheme(this.application.renderer, themePath, options));
 } else {
      this.application.logger.log('To generate markdown please set option --theme markdown');
    }
  }

}
