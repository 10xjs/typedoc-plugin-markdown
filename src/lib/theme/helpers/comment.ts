import * as fs from 'fs';
import * as path from 'path';
import { Reflection } from 'typedoc';
import { MarkdownPlugin } from '../../plugin';
import { relativeUrl } from './relative-url';

const brackets: RegExp = /\[\[([^\]]+)\]\]/g;
const urlPrefix: RegExp = /^(http|ftp)s?:\/\//;
const inlineTag: RegExp = /(?:\[(.+?)\])?\{@(link|linkcode|linkplain)\s+((?:.|\n)+?)\}/gi;

export function comment(this: string) {
  const includePattern: RegExp = /\[\[include:([^\]]+?)\]\]/g;
  const mediaPattern: RegExp = /media:\/\/([^ "\)\]\}]+)/g;

  let text = this;

  /* istanbul ignore next */
  if (MarkdownPlugin.settings && MarkdownPlugin.settings.includes) {
    text = text.replace(includePattern, (match: string, pathString: string) => {
      pathString = path.join(MarkdownPlugin.settings.includes!, pathString.trim());

      if (fs.existsSync(pathString) && fs.statSync(pathString).isFile()) {
        return fs.readFileSync(pathString, 'utf-8');
      }
      return '';
    });
  }

  if (MarkdownPlugin.settings && MarkdownPlugin.settings.media) {
    text = text.replace(mediaPattern, (match: string, pathString: string) => {
      if (fs.existsSync(path.join(MarkdownPlugin.settings.media!, pathString))) {
        return relativeUrl(`media/${pathString}`);
      }
      return match;
    });
  }

  text = replaceInlineTags(replaceBrackets(text));

  return text;
}

function replaceBrackets(text: string): string {
  return text.replace(
    brackets,
    (match: string, content: string): string => {
      const split = splitLinkText(content);
      return buildLink(match, split.target, split.caption);
    },
  );
}

function splitLinkText(text: string): { caption: string; target: string } {
  let splitIndex = text.indexOf('|');
  if (splitIndex === -1) {
    splitIndex = text.search(/\s/);
  }

  if (splitIndex !== -1) {
    return {
      caption: text.substr(splitIndex + 1).replace(/\n+/, ' '),
      target: text.substr(0, splitIndex),
    };
  } else {
    return {
      caption: text,
      target: text,
    };
  }
}

function replaceInlineTags(text: string): string {
  return text.replace(
    inlineTag,
    (match: string, leading: string, tagName: string, content: string): string => {
      const split = splitLinkText(content);
      const target = split.target;
      const caption = leading || split.caption;
      return buildLink(match, target, caption);
    },
  );
}

function buildLink(original: string, target: string, caption: string): string {
  let reflection: Reflection | undefined;
  if (urlPrefix.test(target)) {
    return target;
  }
  if (MarkdownPlugin.reflection) {
    reflection = MarkdownPlugin.reflection.findReflectionByName(target);
  } else if (MarkdownPlugin.project) {
    reflection = MarkdownPlugin.project.findReflectionByName(target);
  }

  if (reflection && reflection.url) {
    if (urlPrefix.test(reflection.url)) {
      target = reflection.url;
    } else {
      target = relativeUrl(reflection.url);
    }
  } else {
    return original;
  }

  return `[${caption}](${target})`;
}
