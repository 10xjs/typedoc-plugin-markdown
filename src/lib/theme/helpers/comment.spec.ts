import { mocked } from 'ts-jest/utils';
import { DeclarationReflection, Reflection } from 'typedoc';
import { Fixture, getFixture, getReflectionByName } from '../../../test/test.utils';
import { MarkdownPlugin } from '../../plugin';
import { MarkdownTheme } from '../theme';
import { comment } from './comment';

jest.mock('../../plugin');
jest.mock('../theme');

let commentsFixture: DeclarationReflection;

describe(`comment helper`, () => {
  beforeAll(() => {
    commentsFixture = getFixture(Fixture.Comment);
    mocked(MarkdownPlugin).theme = {} as MarkdownTheme;
    mocked(MarkdownPlugin).reflection = {
      findReflectionByName: name => {
        return { url: 'reflection-url_.md' };
      },
    } as Reflection;
    mocked(MarkdownPlugin).settings = { includes: './src/test/fixtures/inc/', media: './src/test/fixtures/media/' };
  });

  test(`should compile shortText`, () => {
    const data = getReflectionByName(commentsFixture, 'CommentedClass').comment;
    const result = comment.call(data.shortText);
    expect(result).toMatchSnapshot();
  });

  test(`should compile text`, () => {
    const data = getReflectionByName(commentsFixture, 'CommentedClass').comment;
    const result = comment.call(data.text);
    expect(result).toMatchSnapshot();
  });

  test(`should compile comments from include file`, () => {
    const data = getReflectionByName(getFixture(Fixture.Classes), 'BaseClass').comment;
    const result = comment.call(data.text);
    expect(result).toMatchSnapshot();
  });
});
