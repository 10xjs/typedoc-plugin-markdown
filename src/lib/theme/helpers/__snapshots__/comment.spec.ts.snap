// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comment helper should compile comments from include file 1`] = `
"This is a simple example on how to use BaseClass.

![My image alt text](media/logo.png)

"
`;

exports[`comment helper should compile shortText 1`] = `"A Comment for a class"`;

exports[`comment helper should compile text 1`] = `
"## Some Markup
**with more markup**

A example with decorators that should not parse to tag
\`\`\`
@myDecorator
@FactoryDecorator('a', 'b', 'c')
export class CommentedClass {
  myProp: string = 'myProp';

  @PropDecorator() decoratedProp: string;

  constructor(@ParamDecorator public param: string) { }

  myMethod() { }
}
\`\`\`"
`;
